sklearn/decomposition/_nmf.py

```diff
@@ -1055,13 +1154,20 @@ def non_negative_factorization(X, W=None, H=None, n_components=None, *,
         else:
             W = np.zeros((n_samples, n_components), dtype=X.dtype)
     else:
-        W, H = _initialize_nmf(X, n_components, init=init,
-                               random_state=random_state)
+        if batch_size is None:
```

Comment:
A and B do not have any difference based on the choice of NMF initialization, and are useless if batch_size is None.
I would define them here instead of inside `_initialize_nmf`, and only if batch_size is None.